staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+)
+
+// getCRDSchemaForVersionOrDie returns the validation schema for given version in given CRD.
+func getCRDSchemaForVersionOrDie(crd *apiextensions.CustomResourceDefinition, version string) *apiextensions.CustomResourceValidation {
+	if !hasPerVersionSchema(crd.Spec.Versions) {
+		return crd.Spec.Validation
+	}
+	for _, v := range crd.Spec.Versions {
+		if version != v.Name {
+			continue
+		}
+		return v.Schema
+	}
+	// Given that we call HasServedCRDVersion when serving CRD, this should not happen.
```

Comment:
I think returning an error is better than panicking. Same comment applies in other methods

Reply:
We do have similar cases were we assume things. E.g if default makes sure a pointer id never nil, we don't check for nil repeatedly in the code. I think an error chain for something that cannot happened does not make sense.

Reply:
defaulting is different, since it is run every time we read from etcd or from an incoming request. this is depending on data enforced one time in the past in validation.

Reply:
> this is depending on data enforced one time in the past in validation.

That is not true in this case. It is not validation we depend on, it is a check at the start of each request that make sure the request version exists. So it is at the same level as defaulting which happens every request.